### PR TITLE
[relay][frontend] Return Module from get_workload

### PR DESCRIPTION
**Pull request by zhiics**
This PR makes the relay testing frontend returns `Module` as the frontend parsers also return `Module`.

cc @tqchen @jroesch @wweic @MarisaKirisame

### CODE REVIEW

File: python/tvm/relay/testing/init.py [hunk 1]
```diff
@@ -144,17 +144,16 @@ def create_workload(net, initializer=None, seed=0):
 
     Returns
     -------
-    net : tvm.relay.Function
-        The updated dataflow
+    net : tvm.relay.Module
```
Comment (shoubhik) [inline python/tvm/relay/testing/init.py]: should be `mod` instead of `net`?

**Comment by jroesch** [comment]
Could we add a Python property to `Module` such as `main` so we don't have to write so much boiler plate everywhere?

**Comment by zhiics** [comment]
@jroesch Sure. I will update the rest par of the code for this PR tomorrow.

File: python/tvm/relay/testing/init.py [hunk 2]
```diff
@@ -144,17 +144,16 @@ def create_workload(net, initializer=None, seed=0):
...
 
     params : dict of str to NDArray
         The parameters.
     """
     mod = relay.Module.from_expr(net)
     mod = relay.transform.InferType()(mod)
-    net = mod[mod.entry_func]
     shape_dict = {
-        v.name_hint : v.checked_type for v in net.params}
+        v.name_hint : v.checked_type for v in mod[mod.entry_func].params}
```
Comment (wweic) [inline python/tvm/relay/testing/init.py]:
```suggestion
        v.name_hint : v.checked_type for v in mod.main.params}
```

File: tests/python/frontend/mxnet/test_graph.py [hunk 3]
```diff
@@ -19,66 +19,69 @@
 from tvm.relay import transform
 import model_zoo
 
-def compare_graph(f1, f2):
-    assert relay.analysis.alpha_equal(f1, f2)
+def compare_graph(lhs_mod, rhs_mod):
+    lhs_mod = transform.InferType()(lhs_mod)
+    rhs_mod = transform.InferType()(rhs_mod)
+    assert relay.analysis.alpha_equal(lhs_mod[lhs_mod.entry_func],
```
Comment (wweic) [inline tests/python/frontend/mxnet/test_graph.py]:
```suggestion
    assert relay.analysis.alpha_equal(lhs_mod.main,
```

File: tests/python/frontend/mxnet/test_graph.py [hunk 4]
```diff
@@ -19,66 +19,69 @@
...
+                                      rhs_mod[rhs_mod.entry_func])
```
Comment (wweic) [inline tests/python/frontend/mxnet/test_graph.py]:
```suggestion
                                      rhs_mod.main)
```

File: tests/python/relay/test_autotvm_task_extraction.py [hunk 5]
```diff
@@ -24,46 +24,46 @@ def get_network(name, batch_size):
...
-    return net, params, input_shape
+    return mod, params, input_shape
 
 def test_task_extraction():
     target = 'llvm'
 
-    net, params, input_shape = get_network('resnet-18', batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, input_shape = get_network('resnet-18', batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tests/python/relay/test_autotvm_task_extraction.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tests/python/relay/test_autotvm_task_extraction.py [hunk 6]
```diff
@@ -24,46 +24,46 @@ def get_network(name, batch_size):
...
                                               params=params,
                                               ops=(relay.op.nn.conv2d,))
     assert len(tasks) == 12
 
-    net, params, input_shape = get_network('resnet-18', batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, input_shape = get_network('resnet-18', batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tests/python/relay/test_autotvm_task_extraction.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tests/python/relay/test_autotvm_task_extraction.py [hunk 7]
```diff
@@ -24,46 +24,46 @@ def get_network(name, batch_size):
...
                                             params=params,
                                             ops=(relay.op.nn.dense,))
     assert len(tasks) == 1
 
-    net, params, input_shape = get_network('resnet-18', batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, input_shape = get_network('resnet-18', batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tests/python/relay/test_autotvm_task_extraction.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tests/python/relay/test_autotvm_task_extraction.py [hunk 8]
```diff
@@ -24,46 +24,46 @@ def get_network(name, batch_size):
...
                                             params=params,
                                             ops=(relay.op.nn.conv2d, relay.op.nn.dense))
     assert len(tasks) == 13
 
-    net, params, input_shape = get_network('mobilenet', batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, input_shape = get_network('mobilenet', batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tests/python/relay/test_autotvm_task_extraction.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tests/python/relay/test_autotvm_task_extraction.py [hunk 9]
```diff
@@ -24,46 +24,46 @@ def get_network(name, batch_size):
...
                                             params=params,
                                             ops=(relay.op.nn.conv2d, relay.op.nn.dense))
     assert len(tasks) == 20
 
-    net, params, input_shape = get_network('dcgan', batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, input_shape = get_network('dcgan', batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tests/python/relay/test_autotvm_task_extraction.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tests/python/unittest/test_graph_tuner_core.py [hunk 10]
```diff
@@ -44,8 +44,8 @@ def _create_data(target, dshape, dtype, layout):
     conv2 = relay.nn.conv2d(conv1, w2, channels=32, kernel_size=(3, 3), padding=(1, 1))
     out = relay.add(conv1, conv2)
     net = relay.Function(relay.analysis.free_vars(out), out)
-    net, params = relay.testing.create_workload(net)
-    tasks = autotvm.task.extract_from_program(net,
+    mod, params = relay.testing.create_workload(net)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func],
```
Comment (wweic) [inline tests/python/unittest/test_graph_tuner_core.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main,
```

File: tutorials/autotvm/tune_relay_arm.py [hunk 11]
```diff
@@ -316,10 +317,10 @@ def tune_tasks(tasks,
 def tune_and_evaluate(tuning_opt):
     # extract workloads from relay program
     print("Extract tasks...")
-    net, params, input_shape, _ = get_network(network, batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
-                                            params=params,
-                                            ops=(relay.op.nn.conv2d,))
+    mod, params, input_shape, _ = get_network(network, batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tutorials/autotvm/tune_relay_arm.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tutorials/autotvm/tune_relay_cuda.py [hunk 12]
```diff
@@ -218,9 +219,9 @@ def tune_tasks(tasks,
 def tune_and_evaluate(tuning_opt):
     # extract workloads from relay program
     print("Extract tasks...")
-    net, params, input_shape, out_shape = get_network(network, batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target,
-                                            params=params, ops=(relay.op.nn.conv2d,))
+    mod, params, input_shape, out_shape = get_network(network, batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tutorials/autotvm/tune_relay_cuda.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tutorials/autotvm/tune_relay_mobile_gpu.py [hunk 13]
```diff
@@ -300,8 +301,10 @@ def tune_tasks(tasks,
 def tune_and_evaluate(tuning_opt):
     # extract workloads from relay program
     print("Extract tasks...")
-    net, params, input_shape, _ = get_network(network, batch_size=1)
-    tasks = autotvm.task.extract_from_program(net, target=target, target_host=target_host,
+    mod, params, input_shape, _ = get_network(network, batch_size=1)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func],
```
Comment (wweic) [inline tutorials/autotvm/tune_relay_mobile_gpu.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main,
```

File: tutorials/autotvm/tune_relay_x86.py [hunk 14]
```diff
@@ -177,21 +178,21 @@ def tune_graph(graph, dshape, records, opt_sch_file, use_DP=True):
 def tune_and_evaluate(tuning_opt):
     # extract workloads from relay program
     print("Extract tasks...")
-    net, params, data_shape, out_shape = get_network(model_name, batch_size)
-    tasks = autotvm.task.extract_from_program(net, target=target,
+    mod, params, data_shape, out_shape = get_network(model_name, batch_size)
+    tasks = autotvm.task.extract_from_program(mod[mod.entry_func], target=target,
```
Comment (wweic) [inline tutorials/autotvm/tune_relay_x86.py]:
```suggestion
    tasks = autotvm.task.extract_from_program(mod.main, target=target,
```

File: tutorials/autotvm/tune_relay_x86.py [hunk 15]
```diff
@@ -177,21 +178,21 @@ def tune_graph(graph, dshape, records, opt_sch_file, use_DP=True):
...
                                               params=params, ops=(relay.op.nn.conv2d,))
 
     # run tuning tasks
     print("Tuning...")
     tune_kernels(tasks, **tuning_opt)
-    tune_graph(net, data_shape, log_file, graph_opt_sch_file)
+    tune_graph(mod[mod.entry_func], data_shape, log_file, graph_opt_sch_file)
```
Comment (wweic) [inline tutorials/autotvm/tune_relay_x86.py]:
```suggestion
    tune_graph(mod.main, data_shape, log_file, graph_opt_sch_file)
```

**Comment by zhiics** [comment]
@wweic Thanks. I just refactored part of the code last night. Now everything should be updated to `main`

**Comment by tqchen** [comment]
@zhiics @jroesch I feel we should avoid introducing the additional field mod.main, but instead, directly use mod["main"], so it is no different from getting other functions (remove one level of abstraction). We could likely remove mod.entry_func later.

**Comment by zhiics** [comment]
@jroesch do you have any comments? I am okay with either way. I can refactor to `mod["main"]` if you are okay as well.

**Comment by jroesch** [comment]
Tianqi and I discussed some more, he argues we should just remove and make it so there is no special main function, and we just use main as convention but allow the specification of multiple entry points in the future to passes that do liveness, etc.

**Comment by zhiics** [comment]
@jroesch No worries.

I have one question about the updating the function for a globalvar in a module. 

For example, we may create a module (`mod`) using `from_expr` where a `main` is created or directly create a module using `module({"main": func})`.  Now if I have `mod["main"] = func1`, should I update the function for `main`, or should I keep both? I think we should just check if `main` is in the global list, if so, we then just update the function, right?

**Comment by tqchen** [comment]
@zhiics I think we should update the "main" just as if it is a normal function

**Comment by jroesch** [comment]
I think we should just treat all functions uniformly, there are a couple places where we might need to specify an entry point, and we can just default to main.

**Pull request merged**